### PR TITLE
Feature/admin sidecar

### PR DESCRIPTION
**Pull request by jcnelson**
This PR adds an "administrative side-car" -- an HTTP server that can read and write a Gaia hub's config file and run a command to tell it to restart/reload.

You interact with the side car by POST-ing new configuration settings to it.  This will update the Gaia hub's config on disk.  You make the changes take effect by telling the side-car to restart the Gaia hub.
The side-car also allows the user to query various configuration settings.  All access is mediated by a secret API key defined in the side-car's config file.

# Usage 

**EDIT**:  This documentation below is obsolete, and superseded by the new README.md file in the `hub/admin/` directory of this feature branch.

Sample config file for the admin side-car:
```
$ cat /tmp/gaia-admin.json
{
   "argsTransport": {
    "level": "debug",
    "handleExceptions": true,
    "timestamp": true,
    "stringify": true,
    "colorize": true,
    "json": true
  },
  "regtest": false,
  "testnet": false,
  "port": 8009,
  "apiKeys": ["hello"],
  "gaiaSettings": {
    "configPath": "/tmp/gaia-config.json"
  },
  "reloadSettings": {
    "command": "/bin/sh",
    "argv": ["-c", "echo 'hello world' > /tmp/reload-test"],
    "env": {},
    "setuid": 1000,
    "setgid": 1000
  }
}
```

Build and run with:
```
$ cd admin && npm i && npm run build
$ node lib/index.js /tmp/gaia-admin.json
```

Sample Gaia config to be managed by the side-car:

```
$ cat /tmp/gaia-config.json
{
  "servername": "localhost",
  "port": 4000,
  "driver": "disk",
  "readURL": "http://localhost:4001/",
  "proofsConfig": {
    "proofsRequired": 0
  },
  "diskSettings": {
    "storageRootDirectory": "/tmp/gaia-disk"
  },
  "pageSize": 20,
  "argsTransport": {
    "level": "debug",
    "handleExceptions": true,
    "stringify": true,
    "timestamp": true,
    "colorize": false,
    "json": true
  },
}
```

## Reloading
```
$ cat /tmp/reload-test
cat: /tmp/reload-test: No such file or directory
$ curl -H 'Authorization: bearer hello' -X POST http://localhost:8009/v1/admin/reload
{"result":"OK"}
$ cat /tmp/reload-test
hello world
```

## Get/Set Gaia Settings
```
$ curl -H 'Authorization: bearer hello' http://localhost:8009/v1/admin/gaia
{"config":{"port":4000,"proofsConfig":{"proofsRequired":0}}}
$ curl -H 'Authorization: bearer hello' -X POST --data-raw '{"port": 3001}' http://localhost:8009/v1/admin/gaia
{"message":"Config updated -- you should reload your Gaia hub now."}
$ curl -H 'Authorization: bearer hello' http://localhost:8009/v1/admin/gaia
{"config":{"port":3001,"proofsConfig":{"proofsRequired":0}}}
```

## Get/Set Driver Settings
```
$ curl -H 'Authorization: bearer hello' http://localhost:8009/v1/admin/driver
{"config":{"driver":"disk","readURL":"http://localhost:4001/","pageSize":20,"diskSettings":{"storageRootDirectory":"/tmp/gaia-disk"}}}
$ curl -H 'Authorization: bearer hello' -H 'Content-Type: application/json' -X POST --data-raw '{"driver": "aws", "awsCredentials": {"accessKeyId": "asdf", "secretAccessKey": "jkl;"}}' http://localhost:8009/v1/admin/driver
{"message":"Config updated -- you should reload your Gaia hub now."}
$ curl -H 'Authorization: bearer hello' http://localhost:8009/v1/admin/driver
{"config":{"driver":"aws","readURL":"http://localhost:4001/","pageSize":20,"diskSettings":{"storageRootDirectory":"/tmp/gaia-disk"},"awsCredentials":{"accessKeyId":"asdf","secretAccessKey":"jkl;"}}}
```

## Get/Set Whitelist
```
$ curl -H 'Authorization: bearer hello' http://localhost:8009/v1/admin/whitelist
{"config":{}}
$ curl -H 'Authorization: bearer hello' -H 'Content-Type: application/json' -X POST --data-raw '["15hUKXg1URbQsmaEHKFV2vP9kCeCsT8gUu"]' http://localhost:8009/v1/admin/whitelist
{"message":"Config updated -- you should reload your Gaia hub now."}
$ curl -H 'Authorization: bearer hello' http://localhost:8009/v1/admin/whitelist
{"config":{"whitelist":["15hUKXg1URbQsmaEHKFV2vP9kCeCsT8gUu"]}}
```

### CODE REVIEW

**Comment by kantai** [comment]
Cool, thanks for this @jcnelson 

I still need to spend some time reviewing the code and testing this, but in the meantime, I think this needs a README file in the `admin` folder with instructions on (1) typical usage, (2) an API spec.

**Comment by jcnelson** [comment]
Added the README: https://github.com/blockstack/gaia/tree/feature/admin-sidecar/admin

File: admin/README.md [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+   "argsTransport": {
+    "level": "debug",
+    "handleExceptions": true,
+    "timestamp": true,
+    "stringify": true,
+    "colorize": true,
+    "json": true
+  },
+  "regtest": false,
+  "testnet": false,
```
Comment (kantai) [inline admin/README.md]: what do the testnet/regtest configuration options do?

**Comment by kantai** [comment]
Okay -- after looking this over, I think this API could probably be something a little simpler --

This really only needs to be 3 endpoints:

```
GET /config -> returns the current JSON file
POST /config -> modifies the current JSON file, doing a deep merge of the two objects
POST /reload -> execute the reload command
```

That makes this sidecar pretty generic, so that it doesn't really need to be touched going forward, even if there are changes to the config file schema, etc.

File: admin/README.md (hunk 1 above)
Comment (jcnelson) [reply admin/README.md]: Nothing yet.

**Comment by jcnelson** [comment]
I can do that.  The reasons I separated the endpoints into orthogonal configuration aspects were two-fold:

* Make it so different aspects of the config file could be subject to different authentication schemes -- for example, in shared Gaia hub settings I might want to allow GETs to the whitelist to all users but deny GETs to the driver configuration to all but the admin.  This could be enforced by an nginx reverse proxy.
* Make it hard for a front-end to this API to accidentally modify unrelated configuration aspects -- for example, a client that updates your driver settings should not be able to alter your whitelist.

Maybe these concerns are overblown?  Maybe the better solution would be to have clients simply keep backups of the hub's configuration so you can restore your Gaia hub to a known-good state if something goes wrong.

**Comment by kantai** [comment]
> Make it so different aspects of the config file could be subject to different authentication schemes -- for example, in shared Gaia hub settings I might want to allow GETs to the whitelist to all users but deny GETs to the driver configuration to all but the admin. This could be enforced by an nginx reverse proxy.

I see -- this could be useful. I think my preference here is slightly in favor of just going with generic interfaces to start with, and then adding specific ones later if that use case turns out to be important.

> Make it hard for a front-end to this API to accidentally modify unrelated configuration aspects -- for example, a client that updates your driver settings should not be able to alter your whitelist.

I'm less concerned about this one -- I think that client-side backups is the better solution there (as a client would probably want to do that anyways).

**Comment by jcnelson** [comment]
Made the recommended changes.  The test instructions are now outdated, now that the API endpoints have changed.  Please see the README.md for the latest API documentation.

**Comment by kantai** [comment]
Okay -- cool, this works for me, though I'm not sure if the behavior of `patchConfigFile` is intended:

`Object.assign` does a "shallow merge", e.g.:

```
a = { foo: { baz : 1 } }
b = { foo: { bar : 1 } }
Object.assign(a, b)
// a is now equal to { foo: { bar: 1 } }
```

Which means if I post, e.g.,

```
{ azCredentials: { accountKey: 'bar' } }
``

to change credentials, it will delete the accountName field.

The env var logic in the gaia hub does a "deep merge" to try and patch these things separately, but I'm not sure if that's really necessary here. I'm fine with the behavior as is, but just wanted to make sure that this is intended.

**Comment by jcnelson** [comment]
@kantai The code currently matches the expected behavior -- it treats a Gaia config file's JSON as a set of key/value pairs, and `patchConfigFile` will merge a new set of key/value pairs into the config file's existing key/value pairs.  No deep merge is required.  I have updated the method documentation and test cases to explicitly reflect this design choice.  Thanks for catching that!

File: admin/src/server.js [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+    return Promise.resolve().then(() => {
+      const configPath = this.config.gaiaSettings.configPath
+      return patchConfigFile(configPath, newFields)
+    })
+    .then(() => {
+      const ret = { 
+        statusCode: 200, 
+        status: { 
+          message: 'Config updated -- you should reload your Gaia hub now.'
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply admin/src/server.js]: Yes -- depending on the client implementation, the client can modify the config file multiple times before needing to prompt the user to reload.  For example, the client can have separate pages for updating the driver settings and updating the social proof settings.  Only one reload is needed for changes in both pages to take effect; the user shouldn't have to reload each time a change is made (ideally, the user only reloads when they leave the page, or when they're done making their modifications).

File: admin/src/server.js [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+  handleGetWhitelist(): Promise<{ status: Object, statusCode: number }> {
+    return this.handleGetFields(['whitelist'])
+  }
+
+  handleSetWhitelist(whitelist: Array<string>): Promise<{ status: Object, statusCode: number }> {
+    const ajv = new Ajv()
+    const valid = ajv.validate(WHITELIST_SETTINGS_SCHEMA, whitelist)
+    if (!valid) {
+      logger.error(`Failed to validate whitelist settings: ${JSON.stringify(ajv.errors)}`)
+      const ret = { statusCode: 400, status: { error: 'Invalid whitelist' } }
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply admin/src/server.js]: This code path is outdated.

File: admin/src/server.js [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+  handleGetGaiaSettings(): Promise<{ status: Object, statusCode: number }> {
+    return this.handleGetFields(Object.keys(GAIA_SETTINGS_SCHEMA.properties))
+  }
+  
+  handleSetGaiaSettings(opts: Object): Promise<{ status: Object, statusCode: number }> {
+    const ajv = new Ajv()
+    const valid = ajv.validate(GAIA_SETTINGS_SCHEMA, opts)
+    if (!valid) {
+      logger.error(`Failed to validate Gaia settings: ${JSON.stringify(ajv.errors)}`)
+      const ret = { statusCode: 400, status: { error: 'Invalid Gaia hub configuration' } }
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply admin/src/server.js]: This code path is outdated.

File: admin/test/test.js [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+            env: {
+              "HELLO": "hello world"
+            }
+          }
+        }
+        let server = new AdminAPI(configNoCommand)
+        return server.handleReload()
+      })
+      .then((result) => {
+        t.equal(result.statusCode, 404, 'missing reload command should 404')
```
[the remark this answers is not in the document]
Comment (jcnelson) [reply admin/test/test.js]: Um...these are test warnings. They don't run in production code; only we see them. 

If there is no reload command specified, this endpoint returns 404.  If the command is specified, but fails to run, then this endpoint returns 500.

**Pull request merged**